# Covid19 Global Statistical Data Analysis

First import the modules

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

%matplotlib inline

In [3]:
sns.set_style('darkgrid')

We load the file into a pandas dataframe 

In [4]:
covid_df = pd.read_csv("covid19_global_statistics_2026.csv")

In [5]:
covid_df.head()

,continent,country,population,date,new_cases,active_cases,cases_per_million,total_cases,new_deaths,deaths_per_million,total_deaths,tests_per_million,total_tests
0,Africa,Eritrea,3662244.0,2026-02-16,NaN,0.0,2782.0,10189,NaN,28.0,103.0,6470.0,23693.0
1,Africa,Niger,26083660.0,2026-02-16,NaN,729.0,381.0,9931,NaN,12.0,312.0,9759.0,254538.0
2,Africa,Saint-Helena,6115.0,2026-02-16,NaN,2164.0,354211.0,2166,NaN,NaN,NaN,NaN,NaN
3,South-America,Falkland-Islands,3539.0,2026-02-16,NaN,0.0,545352.0,1930,NaN,NaN,NaN,2439107.0,8632.0
4,North-America,Montserrat,4965.0,2026-02-16,NaN,19.0,282578.0,1403,NaN,1611.0,8.0,3577442.0,17762.0


To get more information about the data set we use `.info()` on the data frame

In [6]:
covid_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 238 entries, 0 to 237
Data columns (total 13 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   continent           236 non-null    object 
 1   country             238 non-null    object 
 2   population          229 non-null    float64
 3   date                238 non-null    object 
 4   new_cases           17 non-null     float64
 5   active_cases        191 non-null    float64
 6   cases_per_million   230 non-null    float64
 7   total_cases         238 non-null    int64  
 8   new_deaths          3 non-null      float64
 9   deaths_per_million  225 non-null    float64
 10  total_deaths        233 non-null    float64
 11  tests_per_million   213 non-null    float64
 12  total_tests         213 non-null    float64
dtypes: float64(9), int64(1), object(3)
memory usage: 24.3+ KB


Count the missing values from the data frame.

In [9]:
covid_df.isna().sum()

continent               2
country                 0
population              9
date                    0
new_cases             221
active_cases           47
cases_per_million       8
total_cases             0
new_deaths            235
deaths_per_million     13
total_deaths            5
tests_per_million      25
total_tests            25
dtype: int64

New Cases and New Deaths have so many empty values, we can simply drop them since they won't be used the much.

In [14]:
for column in covid_df.columns.to_list():
    if column=='new_cases' or column=='new_deaths':
        print('Columns Present')
else:
    print('Columns absent')

Columns absent


In [12]:
covid_df.drop(columns=['new_cases', 'new_deaths'], inplace=True)

In [13]:
for column in covid_df.columns.to_list():
    if column=='new_cases' or column=='new_deaths':
        print('Columns Present')
else:
    print('Columns absent')

Columns absent


In [17]:
covid_df['continent'].unique()

array(['Africa', 'South-America', 'North-America', nan, 'Europe', 'Asia',
       'Oceania', 'All'], dtype=object)

In [19]:
# if I wanted to see how many countries are there in each continent
covid_df.groupby('continent')['country'].count()

continent
Africa           59
All               1
Asia             51
Europe           49
North-America    40
Oceania          21
South-America    15
Name: country, dtype: int64

We can confirm that the data set contains 13 columns and a total of 238 entries.

We get the two values in continent series the are empty(Null)

In [7]:
covid_df[covid_df["continent"].isnull()]

,continent,country,population,date,new_cases,active_cases,cases_per_million,total_cases,new_deaths,deaths_per_million,total_deaths,tests_per_million,total_tests
5,NaN,Diamond-Princess,NaN,2026-02-16,NaN,0.0,NaN,712,NaN,NaN,13.0,NaN,NaN
8,NaN,MS-Zaandam,NaN,2026-02-16,NaN,0.0,NaN,9,NaN,NaN,2.0,NaN,NaN


We can get the values using `iloc[]` to get the index values of the missing continents and assing values to it

In [20]:
covid_df.iloc[5]["continent"]

nan

After obtaining the index we can assing value `Asia` to the space

In [21]:
covid_df.loc[5, "continent"] = "Asia"

In [23]:
covid_df.iloc[5]

continent                         Asia
country               Diamond-Princess
population                         NaN
date                        2026-02-16
active_cases                       0.0
cases_per_million                  NaN
total_cases                        712
deaths_per_million                 NaN
total_deaths                      13.0
tests_per_million                  NaN
total_tests                        NaN
Name: 5, dtype: object

## Working with dates

In [25]:
covid_df["date"].head()

0    2026-02-16
1    2026-02-16
2    2026-02-16
3    2026-02-16
4    2026-02-16
Name: date, dtype: object

The data type of date is currently `object`, so Pandas does not know that this column is a date.

We can convert it into a `datetime` column using the `pd.to_datetime` method.

In [26]:
covid_df["date"] = pd.to_datetime(covid_df.date)
covid_df["date"]

0     2026-02-16
1     2026-02-16
2     2026-02-16
3     2026-02-16
4     2026-02-16
         ...    
233   2026-02-16
234   2026-02-16
235   2026-02-16
236   2026-02-16
237   2026-02-16
Name: date, Length: 238, dtype: datetime64[ns]

In [ ]:
covid_df.columns

In [28]:
numeric_df = covid_df[['active_cases', 'cases_per_million', 'total_cases', 'deaths_per_million', 'total_deaths', 'tests_per_million', 'total_tests']]
numeric_df

,active_cases,cases_per_million,total_cases,deaths_per_million,total_deaths,tests_per_million,total_tests
0,0.0,2782.0,10189,28.0,103.0,6470.0,23693.0
1,729.0,381.0,9931,12.0,312.0,9759.0,254538.0
2,2164.0,354211.0,2166,NaN,NaN,NaN,NaN
3,0.0,545352.0,1930,NaN,NaN,2439107.0,8632.0
4,19.0,282578.0,1403,1611.0,8.0,3577442.0,17762.0
...,...,...,...,...,...,...,...
233,0.0,363666.0,24910387,3389.0,232112.0,7628357.0,522526476.0
234,175641.0,165454.0,24124215,2762.0,402756.0,1875095.0,273400000.0
235,NaN,201399.0,17232066,1194.0,102174.0,1902052.0,162743369.0
236,8716.0,454687.0,11853144,937.0,24414.0,3142326.0,81916639.0


In [29]:
numeric_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 238 entries, 0 to 237
Data columns (total 7 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   active_cases        191 non-null    float64
 1   cases_per_million   230 non-null    float64
 2   total_cases         238 non-null    int64  
 3   deaths_per_million  225 non-null    float64
 4   total_deaths        233 non-null    float64
 5   tests_per_million   213 non-null    float64
 6   total_tests         213 non-null    float64
dtypes: float64(6), int64(1)
memory usage: 13.1 KB


We can drop the `new_deaths` column since it contains few to None values

In [ ]:
numeric_df.new_deaths.info()

In [ ]:
numeric_df.drop(columns=["new_deaths"], inplace=True)

If you want to get numeric data with values in new_cases

In [ ]:
numeric_with_new_cases = numeric_df[numeric_df.new_cases.notnull()]

In [ ]:
numeric_with_new_cases.info()

In [ ]:
numeric_with_new_cases['tests_per_million']

Adding the missing value in `tests_per_million`

In [ ]:
numeric_with_new_cases.at[224, 'tests_per_million']

In [ ]:
numeric_with_new_cases.tests_per_million.mean()

In [ ]:
covid_df.loc[223 : 225]['tests_per_million']

We will replace the value with the average of the entire column

In [ ]:
numeric_with_new_cases.loc[224, 'tests_per_million'] = numeric_with_new_cases.tests_per_million.mean()

Now we can confirm that the value was indeed replaced by the average of the column

In [ ]:
numeric_with_new_cases.loc[224, 'tests_per_million']

In [ ]:
numeric_with_new_cases[['active_cases','total_cases']]

We can plot a line graph of active cases with total cases

First we need to retrieve the data

In [ ]:
active_cases_df = numeric_with_new_cases['active_cases']
total_cases_df = numeric_with_new_cases['total_cases']

In [ ]:
numeric_with_new_cases

In [ ]:
time = np.arange(2000, 2012)
time

In [ ]:
active_cases_df.to_list()

In [ ]:
covid_df.info()

In [ ]:
# Finding the first non-empty index in the series
covid_df.new_cases.first_valid_index() 

In [ ]:
covid_df.loc[11 : 14]

I can create a copy of the data frame 

In [ ]:
covid_df_copy = covid_df.copy()

## Analyzing Data from data frames
Q: What is the total number of reported cases and deaths related to Covid19 in the data frame

In [ ]:
covid_df.columns

In [ ]:
total_cases = covid_df.total_cases.sum()
total_deaths = covid_df.total_deaths.sum()

print(f"The number of reported cases is {total_cases:.2f} and the number of reported deaths is {total_deaths:.2f}")

Q: What is the overall death rate (ratio of reported deaths to reported cases)

In [ ]:
death_rate = covid_df.new_deaths.sum() / covid_df.new_cases.sum()

print(f"The overall reported death rate in the World is {death_rate:.2f}")

We can get the values in new cases which are not Null

In [ ]:
new_cases_df = covid_df[covid_df.new_cases > 0]['new_cases']
new_cases_df

Now we can get values from new deaths

In [ ]:
covid_df.new_deaths.values

We can see the values from new deaths are to low to compare in a line graph

In [ ]:
covid_df[['total_deaths', 'total_cases']].info()

In [ ]:
covid_df.info()

In [ ]:
population = covid_df.population
active_cases = covid_df.active_cases
dates = covid_df.date

We can get a scatter plot of population against activate cases across the world in the data frame

In [ ]:
plt.title("Scatter plot of Population against Activate cases")
sns.scatterplot(
    x='population',
    y='active_cases',
    hue=covid_df.continent.to_list(),
    s=100,
    data=covid_df
);

In [ ]:
plt.hist(covid_df.population);

In [ ]:
dates = covid_df.date
population = covid_df.population
active_cases = covid_df.active_cases

In [ ]:
covid_df.population.plot()
covid_df.new_cases.plot();

To view the first or last few rows of data, we can use the `.head` and `.tail` methods.

In [ ]:
covid_df.head()

In [ ]:
covid_df.tail()

In [ ]:
covid_df.new_deaths.sum()

In [ ]:
covid_df.new_cases.sum()

We can find the first index that doesn't contaion an NaN value using `first_valid_index` method of a series

In [ ]:
covid_df.new_cases.first_valid_index()

We can also check the range of values to check if indeed `12` is the first valid index in new cases series.

In [ ]:
covid_df.new_cases.loc[10 : 14]

We can confirm that indeed this is the first value. Remember `.loc` includes both indexes passed in.

In [ ]:
covid_df.new_deaths.first_valid_index()

In [ ]:
covid_df.new_deaths.loc[40 : 44]

We can get the values in new deaths with indexes

In [ ]:
new_deaths_df = covid_df[covid_df.new_deaths > 0]
new_deaths_df

We can see that there are a total of `2` deaths in North America and `8` deaths in Europe.

In [ ]:
covid_df.groupby(['continent', "country"])["population"].sum()

In [ ]:
covid_df['continent'].unique()

For now let's remove `new_cases` and `new_deaths` since they are irrelevant.

In [ ]:
covid_df

We can drop the two columns inplace, meaning a new data frame won't be created.

In [ ]:
covid_df.drop(columns=['new_cases', 'new_deaths'], inplace=True)

Now we can see that the two columns have been removed.

In [ ]:
covid_df.info()

In [ ]:
covid_df.deaths_per_million.plot()

In [ ]:
plt.plot(covid_df.total_deaths)
plt.plot(covid_df.deaths_per_million)
plt.legend(['total deaths', 'deaths per million']);

In [ ]:
plt.plot(covid_df.total_deaths);

In [ ]:
covid_df.columns

In [ ]:
plt.title('Scatter plot of total cases and cases per million across the World')
sns.scatterplot(x='total_cases', y='cases_per_million', data=covid_df, s=100, hue=covid_df.continent);

In [ ]:
covid_df.columns

In [ ]:
result_df = covid_df[['continent', 'country', 'population', 'active_cases', 'total_tests', 'total_deaths']]

In [ ]:
result_df

In [ ]:
result_df.to_csv('results.csv', index=None)

`index=None` means "don't write row labels (index) to the file" — usually used when saving a DataFrame to a CSV or Excel file.

index=None (or index=False) is important because it prevents an extra unwanted column from being saved to your file.